src/oc_erchef/apps/oc_chef_wm/src/chef_wm_actions_queue_monitoring.erl

```diff
@@ -0,0 +1,473 @@
+         terminate/2,
+         code_change/3,
+         stop/0,
+         status/0,
+         is_queue_at_capacity/0,
+         message_dropped/0,
+         create_pool/0,
+         delete_pool/0,
+         get_pool_configs/0,
+         override_queue_at_capacity/1,
+         sync_check_current_state/0,
+         start_timer/0,
+         stop_timer/0
+
+        ]).
+
+-behaviour(gen_server).
+-define(SERVER, ?MODULE).
+
+
+-record(queue_monitor_state, {
+                % has the max_length of the queue been reached?
+                queue_at_capacity = false,
+                % timer to check max and current queue length
+                timer = undefined,
+                % maximum queue length set by rabbitmq policy
+                % mostly static, but a user *can* change the value
+                % via rabbitmqctl
+                max_length = 0,
+                % last recorded length of the analytics queue
+                last_recorded_length = 0,
+                % number of messages that have NOT been sent to the analytics
+                % queue due to queue_at_capacity = true.
+                % This metric is manually reported by users of the queue
+                % monitor
+                dropped_since_last_check = 0,
+                % monotonically increasing number of messages that have NOT
+                % been sent to the analytics queue due to queue_at_capacity =
+                % true. This number is never reset.
+                total_dropped = 0,
+                % The async worker pid responsible for checking max length and
+                % current length of the queue. There can ONLY be ONE worker
+                % process.
+                worker_process = undefined,
+                % if sync_check_current_state/0 is called, this is the Pid of the
+                % calling process. This is recorded because checking the queue
+                % length and max is async via the worker_process, and the
+                % reply is sent back to this Pid via gen_server:reply()
+                sync_response_process = undefined
+               }).
+
+% exchange is /analytics, so encode the / as %2F
+% NOTE: oc_httpc client is configured to prepend /api
+-define(MAX_LENGTH_PATH, "/policies/%2Fanalytics/max_length").
+-define(QUEUE_LENGTH_PATH, "/queues/%2Fanalytics").
+-define(LOG_THRESHOLD, 0.8).
+-define(POOLNAME, rabbitmq_management_service).
+
+
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+-spec status() -> [{atom(), _}].
+status() ->
+    gen_server:call(?SERVER, status).
+
+% is current queue length >= max queue length?
+% Note, this simply checks the gen_server state, and does NOT
+% ping the RabbitMQ management console.
+% Pinging the management console is handled by an internal timer
+% so we never block waiting for an HTTP response.
+is_queue_at_capacity() ->
+    Timeout = envy:get(oc_chef_wm, rabbitmq_queue_length_timeout_millis, 5000, non_neg_integer),
```

Comment:
A couple of concerns with the timeout, the conditions that could lead up to it, and the effects it'll  have if we start to hit it. I'll write them up shortly in a standalone comment.